engine/src/flutter/shell/platform/common/windowing.h

```diff
@@ -0,0 +1,105 @@
+using FlutterViewId = int64_t;
+
+// Types of windows.
+enum class WindowArchetype {
+  // Regular top-level window.
+  kRegular,
+};
+
+// Possible states a window can be in.
+enum class WindowState {
+  // Normal state, neither maximized, nor minimized.
+  kRestored,
+  // Maximized, occupying the full screen but still showing the system UI.
+  kMaximized,
+  // Minimized and not visible on the screen.
+  kMinimized,
+};
+
+// Converts a |flutter::WindowState| to the string representation of
+// |WindowState| as defined in the framework.
+inline std::string WindowStateToString(WindowState state) {
+  switch (state) {
+    case WindowState::kRestored:
+      return "WindowState.restored";
+    case WindowState::kMaximized:
+      return "WindowState.maximized";
+    case WindowState::kMinimized:
+      return "WindowState.minimized";
+    default:
+      FML_UNREACHABLE();
+  }
+}
+
+// Converts the string representation of |WindowState| defined in the framework
+// to a |flutter::WindowState|. Returns std::nullopt if the given string is
+// invalid.
+inline std::optional<WindowState> StringToWindowState(std::string_view str) {
+  if (str == "WindowState.restored")
+    return WindowState::kRestored;
+  if (str == "WindowState.maximized")
+    return WindowState::kMaximized;
+  if (str == "WindowState.minimized")
+    return WindowState::kMinimized;
+  return std::nullopt;
+}
+
+// Settings used for creating a Flutter window.
+struct WindowCreationSettings {
+  // Type of the window.
+  WindowArchetype archetype = WindowArchetype::kRegular;
+  // Requested size of the window's client area, in logical coordinates.
+  Size size;
+  // Minimum size of the window's client area, in logical coordinates.
+  std::optional<Size> min_size;
+  // Maximum size of the window's client area, in logical coordinates.
+  std::optional<Size> max_size;
+  // Window title.
+  std::optional<std::string> title;
+  // Initial state of the window.
+  std::optional<WindowState> state;
+};
+
+// Settings for modifying a Flutter window.
+struct WindowModificationSettings {
+  // The new requested size, in logical coordinates.
+  std::optional<Size> size;
+  // The new window title.
+  std::optional<std::string> title;
+  // The new window state.
+  std::optional<WindowState> state;
+};
+
+// Window metadata returned as the result of creating a Flutter window.
+struct WindowMetadata {
+  // The ID of the view used for this window, which is unique to each window.
+  FlutterViewId view_id = 0;
+  // The type of the window.
+  WindowArchetype archetype = WindowArchetype::kRegular;
+  // Size of the created window, in logical coordinates.
+  Size size;
+  // The ID of the view used by the parent window. If not set, the window is
+  // assumed a top-level window.
+  std::optional<FlutterViewId> parent_id;
+  // The initial state of the window, or std::nullopt if not a regular window.
+  std::optional<WindowState> state;
+};
+
+}  // namespace flutter
```

Comment:
I don't know how I feel about this. Especially when looking at later PRs, which seem to put quite a bit of additional logic here, like [positioning for popovers](https://github.com/canonical/flutter/blob/popup-menu/engine/src/flutter/shell/platform/common/windowing.cc). Yes, we share some code between embedders already, but this seems very different as it implements a specific algorithm.

We have formal embedder API, "informal" platform channels, and now we have algorithms that embedders are expected to implement, placed at arbitrary location in the engine. What about out of tree embedders?

Wouldn't it be better if something like [this](https://github.com/canonical/flutter/blob/popup-menu/engine/src/flutter/shell/platform/common/windowing.cc) was implemented in dart instead?

Reply:
In the case of positioning, modern compositors do this themselves, as an application never has all the context required to do this correctly (or is not allowed the context to maintain application isolation). For compositors that do not do this, the application needs to perform this, thus each compositor needs to implement this logic.

The difficulty in making shared dart code is the embedder is already running native code when it needs it. If we were to make this code run before requests reach the embedder it would make the API more complicated. In all cases, the shared code is more of a set of helper functions, as the behaviour always depends on the compositor in use.

Reply:
I don't quite understand how this addresses my concern. There is no precedent for sharing logic between embedders in-tree like this and I think this definitely needs further discussion.

I have proposal for synchronous ffi-based API https://docs.flutter.dev/go/multi-window-api-ffi. With synchronous API we can easily move all this to Dart code so that it can be accessed by all embedders, not just the "blessed" ones.

Reply:
Is this not similar to the shared text input model code?

Reply:
Very vaguely. TextInputModel doesn't implement particular algorithm, at least not in a way where all embedders rely on. For example on macOS we're only using a tiny subset of it. iOS or Android is not using it at all. So it's more of a utility class rather than defining logic that embedders are supposed to follow.

Also text input is one of the most problematic parts of Flutter, and if we were starting from scratch we would be likely using a very different approach, see https://github.com/flutter/flutter/issues/150460. So modeling new code after text input doesn't seem like the best idea.

Reply:
That's actually how this is meant to be used - as helpers - depending on platform, the embedder might use e.g. the positioner information itself (like on Windows) - on others (like Wayland) that context will be passed on verbatim to the compositor, so there's no point in making those calculations in Flutter at all. On others yet, child windows might not be supported altogether (mobile platforms, automotive).

The implementation here will only be shared between platforms that can meaningfully make use of it (floating window desktops - which sure, are the majority of them, but not all). Other platforms should default to not supporting child windows until/unless they grow support for at least a subset of them.